# Examining an Identified Accession for Presence of a TSL4-rated transcript

This takes an accession identified previously in the last notebook of this series,[Selecting Logan Search results by Sequencing Technology using Python in Jupyter](Selecting_Logan_Search_results_by_Sequencing_Technology_using_Python_in_Jupyter.ipynb), and follows-up to see if a representative of the TSL4-rated transcript ENST00000567329 (USP7-217) can be identified since the query results suggests its presence.

**Two important notes about this Jupyter Notebook**
- Unlike the prior notebooks in this series, this one is not meant to be run in this session. It has already been run, and so you can review it by scrolling down.

  If you did want to run it: This was developed in a blast-binder session launched from [my blast-binder repo](https://github.com/fomightez/blast-binder) (note that also has PatMatch installed, but that software wasn't used here — although earlier I checked some preliminary results using it and have noted that here).  
  Follow the directions below to launch yet another mybinder-served session and bring this notebook over to that session.
  
- it is a rough sketch of an analysis at this stage

  There's a lot of USP7-related transcripts and this is only rough classification that doesn't take this into account. It is done to simply give a general idea of the TSL4-rated transcript relative other USP7 transcripts.


**IMPORTANT- STEP-BY-STEP HOW TO RUN THIS NOTEBOOK, IF YOU NEED TO:**  

If you are reading this in an active session launched from [the 'logan_results_analysis-binder'](https://github.com/fomightez/logan_results_analysis-binder) as intended, you should be able to review this notebook pre-run and examine all the results. However, if you did want to run it.
As mentioned above, this needs to be run in a different session with BLAST software available. This is to keep the current environment less complex at this time as only this last notebook needs that.
- Right-click on the following link and choose to open it in a new browser window so that you can continue to read this: [click this](https://mybinder.org/v2/gh/fomightez/blast-binder/master?urlpath=%2Flab%2Ftree)
- When the session spins up, open a new notebook backed by a Python kernel by clicking on the upper left tile in the main Jupyter pane and then use copy-and-paste to run the following command in it:

```shell
!curl -OL https://raw.githubusercontent.com/fomightez/logan_results_analysis-binder/refs/heads/main/notebooks/Examining_an_Indentified_Accession_for_Presence_of_a_TSL4-rated_transcript.ipynb
```

- After a few seconds the notebook `Examining_an_Indentified_Accession_for_Presence_of_a_TSL4-rated_transcript.ipynb` should show up in the file browser pane on the left. Now you are ready to run that.

- Double-click on the notebook `Examining_an_Indentified_Accession_for_Presence_of_a_TSL4-rated_transcript.ipynb` to open it.

- When the notebook opens start running the sections below this point and continue on using the notebook there. Importantly, if you modify `Examining_an_Indentified_Accession_for_Presence_of_a_TSL4-rated_transcript.ipynb` and make soemthing useful, be sure to download and keep that version as this one is no way connected to it.

Actual content continues below...

-----

### Preparation for Analysis STAGE 1, Looking into identified accession & mapping its reads

In the previous notebook in this series, [Selecting Logan Search results by Sequencing Technology using Python in Jupyter](Selecting_Logan_Search_results_by_Sequencing_Technology_using_Python_in_Jupyter.ipynb), one of the accessions with long reads identified was `SRR23849628`. 

The Logan Search data does include a substantial amount of metadata and examining the row in that notebook might show tissue source of the data. In the case of `SRR23849628`, I don't see that as the case.  
However, we need additional metadata not there, such as URLs to retrieve the data. And it is also nice to know how to programmatically access the metadata, and so we'll use the packages `ffq` and `bio` here to access such metata.

More about the two packages, with additional demonstrations can be found [here](https://www.biostars.org/p/9522636/#) and in [my demonstration notebook, 'demo ffq for finding sequencing data and metadata from public databases'](https://nbviewer.org/github/fomightez/cl_sq_demo-binder/blob/master/notebooks/demo%20ffq%20for%20finding%20sequencing%20data%20and%20metadata%20from%20public%20databases.ipynb), found in [my cl_sq_demo-binder repo](https://github.com/fomightez/cl_sq_demo-binder).

Let's first run the next cell to install the packages.

In [1]:
%pip install ffq bio -q

Note: you may need to restart the kernel to use updated packages.


####  Analysis STAGE 1, Looking into identified accession & mapping its reads

With the preparaton out of the way, we can use the packages to look at the metadata.

In [2]:
!ffq SRR23849628

[2025-04-17 20:14:05,029]    INFO Parsing run SRR23849628
{
    "SRR23849628": {
        "accession": "SRR23849628",
        "experiment": "SRX19662564",
        "study": "SRP410260",
        "sample": "SRS17033009",
        "title": "PromethION sequencing; GSM7093690: 35cycle_10X; Homo sapiens; RNA-Seq",
        "attributes": {
            "ENA-FIRST-PUBLIC": "2023-06-23",
            "ENA-LAST-UPDATE": "2023-06-23"
        },
        "files": {
            "ftp": [
                {
                    "accession": "SRR23849628",
                    "filename": "SRR23849628_1.fastq.gz",
                    "filetype": "fastq",
                    "filesize": 10790875511,
                    "filenumber": 1,
                    "md5": "05b77f9a3d01bd63e66b796c16e86b90",
                    "urltype": "ftp",
                    "url": "ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR238/028/SRR23849628/SRR23849628_1.fastq.gz"
                }
            ],
            "aws": [],
            "g

Note, often there will be `aws` and `ncbi` URLs. (It seems ones deposited at the UK ebi aren't propagated to the other resources. And so options are more limited than you'll typically observe.)

I've noted with `ffq`, you get more associated metadata if you use the 'sample' listed for the query (not the same as the 'BioSample' that begins usually with `SAMN`). The one above shows `SRS17033009` for 'sample' and so let's run that, too:

In [3]:
!ffq SRS17033009

[2025-04-17 20:14:06,938]    INFO Parsing sample SRS17033009
[2025-04-17 20:14:07,105] WARNING Failed to parse sample information from ENA XML. Falling back to ENA search...
[2025-04-17 20:14:07,243]    INFO Getting Experiment for SRS17033009
[2025-04-17 20:14:07,243]    INFO Parsing Experiment SRX19662564
[2025-04-17 20:14:07,403] WARNING Failed to parse experiment information from ENA XML. Falling back to ENA search...
[2025-04-17 20:14:07,540] WARNING There is 1 run for SRX19662564
[2025-04-17 20:14:07,540]    INFO Parsing run SRR23849628
{
    "SRS17033009": {
        "accession": "SRS17033009",
        "title": "35cycle_10X",
        "organism": "Homo sapiens",
        "attributes": {
            "collection_date": "missing",
            "treatment": "35cycles",
            "INSDC secondary accession": "SRS17033009",
            "NCBI submission package": "Generic.1.0",
            "organism": "Homo sapiens",
            "geo_loc_name": "missing",
            "cell type": "Myeloma

That indicates the source of this sequence data is: 'Myeloma cell lines'.

Let's do this with `bio` as well.

In [4]:
!bio search SRR23849628

[
    {
        "run_accession": "SRR23849628",
        "sample_accession": "SAMN33743856",
        "sample_alias": "GSM7093690",
        "sample_description": "35cycle_10X",
        "first_public": "2023-06-23",
        "country": "missing",
        "scientific_name": "Homo sapiens",
        "fastq_bytes": "10790875511",
        "base_count": "10715195111",
        "read_count": "6444874",
        "library_name": "GSM7093690",
        "library_strategy": "RNA-Seq",
        "library_source": "TRANSCRIPTOMIC",
        "library_layout": "SINGLE",
        "instrument_platform": "OXFORD_NANOPORE",
        "instrument_model": "PromethION",
        "study_title": "Counting and correcting errors within unique molecular identifiers to generate absolute numbers of sequencing molecules [scRNA-Seq]",
        "fastq_url": [
            "https://ftp.sra.ebi.ac.uk/vol1/fastq/SRR238/028/SRR23849628/SRR23849628_1.fastq.gz"
        ],
        "info": "11 GB files; 6.4 million reads; 10715.2 million seq

This reports number of reads as 6.4 million.  
Hmmm... in this case `bio search` doesn't give the source of the data as clear as `ffq` does. It is atypical, though as seen from this unrelated example `!bio search SRR17607594` if you care to run it.

Given the difference of the software results with the same identifiers, and ffq being better when using 'sample', it is good to look at a few examples and make sure you follow-up on any dead ends you seem to hit with one software, perhaps with another.

In this case, both give the same URLs for retrieval of the data; however, as I said though `ffq` will typically offer you more options for retrieval. Even in cases where other options exist, `bio search` won't feature the other URLs, at least currently. (You can convince yourself of this running `!ffq ERR5670887` vs. `!bio search ERR5670887`. The data there is not related to here, but illustrates what I am saying.)

With that URL, Minimap2 was used to do alignment with the data in accession `SRR23849628` to the human genome, specifically [Genome Reference Consortium Human GRCh38.p14/hg38](https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz).

I used `samtools view` to limit the data to reads that map to the USP7 locus, plus/minus 10 bps (human chr16:8892087-8975338).

I have supplied the resulting pertinent data & output here for examination.

-----

### Preparation for Analysis STAGE 2, Running this notebook using data collected by mapping reads

Now already have generated `USP7_locus_mapped_SRX19662564_reads.fastq` which came from running minimap with the identified accession `SRR23849628 `. More about that below. We'll just need to bring it to this active session.

Likewise, will need to bring here the already-generated files:
- `mapped_stats_all_reads_$(ENA_EXPERIMENT).tsv`
- `mapped_stats_just_USP7_locus_$(ENA_EXPERIMENT).tsv`
- `Homo_sapiens_ENST00000344836_9_sequence_USP7-MANE_trancript.fasta`
- `Homo_sapiens_ENST00000567329_1_sequence_USP7-217_trancript.fasta`

Run the following cell to get all those and place them where expected in this active session:

In [5]:
!git clone https://github.com/fomightez/logan_results_analysis-binder.git && echo "Cloned successfully...continuing on.." # `&&` parts makes sure it completes this before contuing on to move
!mv logan_results_analysis-binder/notebooks/supporting_tables_and_data/ .
!rm -rf logan_results_analysis-binder

Cloning into 'logan_results_analysis-binder'...
remote: Enumerating objects: 371, done.
remote: Counting objects: 100% (79/79), done.
remote: Compressing objects: 100% (52/52), done.
remote: Total 371 (delta 46), reused 59 (delta 27), pack-reused 292 (from 1)
Receiving objects: 100% (371/371), 40.74 MiB | 23.72 MiB/s, done.
Resolving deltas: 100% (239/239), done.
Cloned successfully...continuing on..
mv: cannot move 'logan_results_analysis-binder/notebooks/supporting_tables_and_data/' to './supporting_tables_and_data': Directory not empty


Run the next cell to install additional packages needed for the following section.

In [6]:
%%capture
%pip install pyfaidx
%conda install -c bioconda fqgrep

Get utility scripts needed.

In [7]:
!curl -O https://raw.githubusercontent.com/fomightez/sequencework/master/blast-utilities/blast_to_df.py

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 13771  100 13771    0     0  75538      0 --:--:-- --:--:-- --:--:-- 75664


------
### Analysis of the Mapped Reads

#### Stats about reads mapped

Let's look at some details about all the reads mapped from that accession and then focus in on the ones stored in `USP7_locus_mapped_SRX19662564_reads.fastq`.

Run the next cell for looking at all:

In [8]:
!cat supporting_tables_and_data/mapped_stats_all_reads_SRX19662564.tsv 

10616211	0	total (QC-passed reads + QC-failed reads)
6444874	0	primary
2738386	0	secondary
1432951	0	supplementary
0	0	duplicates
0	0	primary duplicates
9387725	0	mapped
88.43%	N/A	mapped %
5216388	0	primary mapped
80.94%	N/A	primary mapped %
0	0	paired in sequencing
0	0	read1
0	0	read2
0	0	properly paired
N/A	N/A	properly paired %
0	0	with itself and mate mapped
0	0	singletons
N/A	N/A	singletons %
0	0	with mate mapped to a different chr
0	0	with mate mapped to a different chr (mapQ>=5)


Note the line primary reads number, i.e., `6444874	0	primary`, matches what `!bio search SRR23849628` gave earlier.  
Now for just the locus.

In [9]:
!cat supporting_tables_and_data/mapped_stats_just_USP7_locus_SRX19662564.tsv

1109	0	total (QC-passed reads + QC-failed reads)
863	0	primary
7	0	secondary
239	0	supplementary
0	0	duplicates
0	0	primary duplicates
1109	0	mapped
100.00%	N/A	mapped %
863	0	primary mapped
100.00%	N/A	primary mapped %
0	0	paired in sequencing
0	0	read1
0	0	read2
0	0	properly paired
N/A	N/A	properly paired %
0	0	with itself and mate mapped
0	0	singletons
N/A	N/A	singletons %
0	0	with mate mapped to a different chr
0	0	with mate mapped to a different chr (mapQ>=5)


There's 863 reads mapped to USP7 out of the 6.4 million included in `SRR23849628`.  
Can we see the TSL4-rated one we are interested in and see what percentage it is?


----

#### Analysis of the Mapped Reads for presence of TSL4-rated transcript USP7-217.

We now have reads we can mine for information about the transcript of interest, the USP7-217 retained intron-query sequence ENST00000567329 (USP7-217).

**Using Fulcrum Genomics' `fqgrep` to search for a match to the query sequence among the reads.**


First use fqgrep to quickly look and see if can locate reads with the match. (See [my fqgrep-binder](https://github.com/fomightez/fqgrep-binder) for more details about Fulcrum Genomics' `fqgrep`, including a launchable demonstration notebook.)

In [10]:
USP7_TSL4_RET_I = 'GGCAATAAAACAGTAAATATTGTTAATAGCG'
!fqgrep --reverse-complement --color auto {USP7_TSL4_RET_I} supporting_tables_and_data/USP7_locus_mapped_SRX19662564_reads.fastq

@SRR23849628.5083175
TACGTATTGCTAAGCAGTGGTATCAACGCAGAGTGAATTTTTGCAAAAAACAGATCCTAAGGACCCTGCAAATTATATTCTTCATGCAGTCCTGGTTCATAGTGGAGATAATCATGGTGGACATTATGTGGTTTATCTAAACCCCAAAGGGGATGGCAAATGGTGTAAATTTGATGACGACGTGGTGTCAAGGTGTACTAAAGAGGAAGCAATTGAGCACAATTATGGGGGTCACGATGACGACCTGTCTGTTCGACACTGCACTAATGCTTACATGTTAGTCTACATCAGGGAATCAAAACTGAGTGAAGTTTTACAGGCGGTCACCGACCATGATATTCCTCAGCAGTTGGTGGAGCGATTACAAGAAGAGAAAAGGATCGAGGCTCAGAAGCGGAAGGAGCGGCAGGAAGCCCATCTCTATATGCAAGTGCAGATAGTCGCAGAGGACCAGTTTTGTGGCCACCAAGGGAATGACATGTACGATGAAGAAAAAGTGAAATACACTGTGTTCAAAGTATTGAAGAACTCCTCGCTTGCTGAGTTTGTTCAGAGCCTCTCTCAGACCATGGGATTTCCACAAGATCAAATTCGATTGTGGCCCATGCAAGCAAGGAGTAATGGAACAAAACGACCAGCAATGTTAGATAATGAAGCCGACGGCAATAAAACAGTAAATATTGTTAATAGCGTATCTGGTTTGGAACCGTGCAGAAGGCGTTAGTCCTCTGCACTTAGTGCAGCTTGTTTCCCTTTTGGTCCACTTAACTAGAATTGGACGTTTTCTTCAATACTTGACTGTAGTTTTTCGCTCTGTCACCTAAGCCATTAGACTCTTCTAAAATCAGCGTCTCTTTTGGAAAATAGATGATTGAGCTCAGTGATAATGAAAACCCTTGGACAATATTCCTGGAAACAGTTGATCCCAGCTGGCTGCTAGTGGAGCGACCTTACCCAAGTTTGATAAAGATCGTAAGTG

Two reads have exact matches to the USP7-217 retained intron-query sequence.
Actually no other reads seem to have anything because I checked allowing 3 mismatches in two ways:

**One way I allowed mistmatches**  
One with Indraniel Das' `fqgrep`, go [here](https://github.com/fomightez/indraniel_fqgrep-binder) to run such sessions.

These were the commands there:

```shell
!fqgrep -c -p 'GGCAATAAAACAGTAAATATTGTTAATAGCG' -m 3 ./USP7_locus_mapped_SRX19662564_reads.fastq
!fqgrep -c -p 'CGCTATTAACAATATTTACTGTTTTATTGCC' -m 3 ./USP7_locus_mapped_SRX19662564_reads.fastq
```

(note 'CGCTATTAACAATATTTACTGTTTTATTGCC' is the reverse complement of `GGCAATAAAACAGTAAATATTGTTAATAGCG` (tag `USP7_TSL4_RET_I`) that I used in that manner because in my preliminary investigations of  Indraniel Das' `fqgrep` so far I have not seen a flag to do reverse complement like both Fulcrum Genomics' `fqgrep` and PatMAtch have!!)

**Second way I allowed mistmatches**  
And, second, for sessions launched from repos that have PatMatch specified to be installed in resulting Jupyter sessions, [my patmatch-binder repo](https://github.com/fomightez/patmatch-binder) (or [my blast-binder repo](https://github.com/fomightez/blast-binder) has PatMAtch, too), checked with allowing 3 mismatches:

```python
RETI = "GGCAATAAAACAGTAAATATTGTTAATAGCG"
!perl patmatch_1.2/patmatch.pl -c {RETI} demo.fasta 3 ids
```

For PatMatch use, first had to covert `USP7_locus_mapped_SRX19662564_reads.fastq` to FASTA using AWK.

UPSHOT OF BOTH: still only those two with the perfect matches seen!  
Note that I would suggest doing analyses like these with tools that allow fuzzy searching, too, when considering your own sequences and read data, especially with long read technology that can show errors. I am only focusing just on Fulcrum Genomics' fqgrep because it is easy to install and works in this case since there is a perfect match. However, you'll note I looked into that in additional, separate investigation. 

Let's delve into that second listed read containing the match some. Specifically the one:

```text
@SRR23849628.535194
TGTATTGCTCTACACGACGCTCTTCCGATCTCACCGACCATGATATTCCTCAGCAGTTGGTGGAGCGATTACAAAGAAGAAGAAAAGGATCGAGGCTCAGAAGCGGAAGGAGCGGCAGGAAGCCCATCTCTATATAGTCGCAGAGGACCAGTTTTGTGGCCACCAAGGGAATGACATGTACGATGAAGAAAAAGTGAAATACACTGTGTTCAAAGATATCAAGAAACCCCCTCGCTTGCTGAGTTTGTTCAGAGCCTCTCTCAGACCATGGGATTTCCACAAGATCAAATTCGATTGTGGCCCATGCAAGCAAGGAGTAATGGAACAAACGACCAGCAATGTTAGATAATGAAGCCGACGGCAATAAAACAGTAAATATTGTTAATAGCGTATCTGGTTTGGAACCGTGCAGAAGGCGTTAGTCCTTCTGCACTTAGTGCAGCTTGTTTCCCTTTTGGTCCACTTAACTAGAATTGGACGTTTTCTTCAATACTTGACTTAGTTTTTCGCTCTGTCACCTAAGCCATTAGACTCTTCTAAAATCAGCGTCTCTTTTGGAAATAGATGATTGAGCTCAGTGATAATGAAAACCCTTGGACAATATTCCTGGAAACAGTTGATCCCGAGCTGGCTGCTAGTGGAGCGACCTTACCCAAGTTTGATAAAGATCGTAAGTGCCCACGCAGACGCCTGCCTGCACTTAGAGCAGTAGCGTTGGATTCCTGGATTGTTGTTCTAAACACACAGGGTTAAGCCTTCTTCCCTTGCAGAAAGATGTGTTGATCAATTCCACAAAAAAAAAAAAAAAAAAAAAAAGTCGGCTAAGCTGGGATTTGGCCGTTGATCGGAAGAGCGTCGTGTAGAGCAATACGTAACTGAACGAAGTAACGACAAT
```

(Note YOU HAVE TO SCROLL TO THE RIGHT TO SEE ALL OF IT. Ot doesn't wrap in the notebook automatically. I haven't decided if I should make it have lines instead?)

That second one looks immediately as a possibility as it is small, in the range of the expected few hundred bps.  
So let's start there and use BLAST to compare the entire 895 bp of it to the  567 bp USP7-217 transcript pairwise via the NCBI portal.  That gives:

```text
Query Cover 62% 
Alignment statistics for match #1
Score	Expect	Identities	Gaps	Strand
910 bits(1008)	0.0	548/572(96%)	19/572(3%)	Plus/Plus
Query  56   GTTGGTGGAGCGATTACAAAGAAGAAGAAAAGGATCGAGGCTCAGAAGCGGAAGGAGCGG  115
            ||||||||||||||||||| ||||| ||||||||||||||||||||||||||||||||||
Sbjct  1    GTTGGTGGAGCGATTACAA-GAAGA-GAAAAGGATCGAGGCTCAGAAGCGGAAGGAGCGG  58

Query  116  CAGGAAGCCCATCTCTATAT------------AGTCGCAGAGGACCAGTTTTGTGGCCAC  163
            ||||||||||||||||||||            ||||||||||||||||||||||||||||
Sbjct  59   CAGGAAGCCCATCTCTATATGCAAGTGCAGATAGTCGCAGAGGACCAGTTTTGTGGCCAC  118

Query  164  CAAGGGAATGACATGTACGATGAAGAAAAAGTGAAATACACTGTGTTCAAAGATATCAAG  223
            ||||||||||||||||||||||||||||||||||||||||||||||||||||   | |||
Sbjct  119  CAAGGGAATGACATGTACGATGAAGAAAAAGTGAAATACACTGTGTTCAAAGTATTGAAG  178

Query  224  AAACCCCCTCGCTTGCTGAGTTTGTTCAGAGCCTCTCTCAGACCATGGGATTTCCACAAG  283
            ||  | ||||||||||||||||||||||||||||||||||||||||||||||||||||||
Sbjct  179  AA--CTCCTCGCTTGCTGAGTTTGTTCAGAGCCTCTCTCAGACCATGGGATTTCCACAAG  236

Query  284  ATCAAATTCGATTGTGGCCCATGCAAGCAAGGAGTAATGGAACAAA-CGACCAGCAATGT  342
            |||||||||||||||||||||||||||||||||||||||||||||| |||||||||||||
Sbjct  237  ATCAAATTCGATTGTGGCCCATGCAAGCAAGGAGTAATGGAACAAAACGACCAGCAATGT  296

Query  343  TAGATAATGAAGCCGACGGCAATAAAACAGTAAATATTGTTAATAGCGTATCTGGTTTGG  402
            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
Sbjct  297  TAGATAATGAAGCCGACGGCAATAAAACAGTAAATATTGTTAATAGCGTATCTGGTTTGG  356

Query  403  AACCGTGCAGAAGGCGTTAGTCCTTCTGCACTTAGTGCAGCTTGTTTCCCTTTTGGTCCA  462
            |||||||||||||||||||||||| |||||||||||||||||||||||||||||||||||
Sbjct  357  AACCGTGCAGAAGGCGTTAGTCCT-CTGCACTTAGTGCAGCTTGTTTCCCTTTTGGTCCA  415

Query  463  CTTAACTAGAATTGGACGTTTTCTTCAATACTTGACT-TAGTTTTTCGCTCTGTCACCTA  521
            ||||||||||||||||||||||||||||||||||||| ||||||||||||||||||||||
Sbjct  416  CTTAACTAGAATTGGACGTTTTCTTCAATACTTGACTGTAGTTTTTCGCTCTGTCACCTA  475

Query  522  AGCCATTAGACTCTTCTAAAATCAGCGTCTCTTTTGGAAATAGATGATTGAGCTCAGTGA  581
            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
Sbjct  476  AGCCATTAGACTCTTCTAAAATCAGCGTCTCTTTTGGAAATAGATGATTGAGCTCAGTGA  535

Query  582  TAATGAAAACCCTTGGACAATATTCCTGGAAA  613
            ||||||||||||||||||||||||||||||||
Sbjct  536  TAATGAAAACCCTTGGACAATATTCCTGGAAA  567
```

That read is continguous and well matching to the 567 bp of USP7-217 transcript over a large span that includes two splice junctions, plus importantly the lack of a junction where the retained intron continues to make of the transcript (around the 300 bp point of the USP7-217 transcript on the bottom in the alignment), and so it seems **that read is indeed an example of the TSL4-rated transcript ENST00000567329 (USP7-217)**.

That explains the central 567 bps of the read. In looking into what the rest of the 895 bps is I realized what I Ensembl gave me when I clicked 'Download' for the ENST00000567329 (USP7-217) 'cDNA' only included part of the content you'll see if you go to [here](https://useast.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000187555;r=16:8895045-8956380;t=ENST00000567329) and toggle the 'Exons' view using the navigation panel in the upper, side left of the page. Specifically it starts with `GTTGGTGGAGCGATTACAAGA...` and ends with `...AACCCTTGGACAATATTCCTGGAAA` and that 567 bp sequence Ensembl gives leaves off the part the page shows as `5' upstream sequence` (`.........cctgcaggtgaagttttacaggcggtcaccgaccatgatattcctcagca`) and `3' downstream sequence` (`cagttgatcccgagctggctgctagtggagcgaccttacccaagtttgat.........`). So the `5' upstream sequence` explains the first 55 bps. You can even see the match to what I included above that I copied and pasted from the Ensmbl page. Below on the top is from Enxmbl and on the bottom is the start of the `@SRR23849628.535194` read, the 55bps before the part shown in the BLAST result above:  

```text
cctgcaggtgaa--gttttacaggcgg-----------tcaccgaccatgatattcctcagca
        ||.|  |.|.||||.|..|           |||||||||||||||||||||||||
--------TGTATTGCTCTACACGACGCTCTTCCGATCTCACCGACCATGATATTCCTCAGCA
```

The extreme left end may be adapter/technical sequence, but the main part is clearly a match between what Ensembl indicates as `5' upstream sequence` and the read data.

It is a similar thing for the 3' end. In fact, you probably already noted the Poly(A) tail being clearly defined towards the end of the read @SRR23849628.535194 shown earlier. And what is between the 614th nucleotide of the read and there matches nicely to the `3' downstream sequence`. Here shows on the top what Esembl has for the start of `3' downstream sequence` with the read part after the BLAST pairwise alignment shown on the bottom:

```text
cagttgatcccgagctggctgctagtggagcgaccttacccaagtttgat.........
CAGTTGATCCCGAGCTGGCTGCTAGTGGAGCGACCTTACCCAAGTTTGATAAAGATCGTAAGTGCCCACGCAGACGCCTGCCTGCACTTAGAGCAGTAGCGTTGGATTCCTGGATTGTTGTTCTAAACACACAGGGTTAAGCCTTCTTCCCTTGCAGAAAGATGTGTTGATCAATTCCACAAAAAAAAAAAAAAAAAAAAAAA
```

That matches really well. And in fact what Ensembl represents with `.........` also matches very well as can bee see by taking more of the downstream part of that region of USP7 and aligning with [EMBOSS Needle - pairwise sequence alignment](https://www.ebi.ac.uk/jdispatcher/psa/emboss_needle?stype=dna&matrix=EDNAFULL):

```text
USP73p_001         1 CAGTTGATCCCGAGCTGGCTGCTAGTGGAGCGACCTTACCCAAGTTTGAT     50
                     ||||||||||||||||||||||||||||||||||||||||||||||||||
535194_001         1 CAGTTGATCCCGAGCTGGCTGCTAGTGGAGCGACCTTACCCAAGTTTGAT     50

USP73p_001        51 AAAGATCGTAAGTGCCCACG-AGACGCCTGCCTGCACTTAGAGCAGTAGC     99
                     |||||||||||||||||||| |||||||||||||||||||||||||||||
535194_001        51 AAAGATCGTAAGTGCCCACGCAGACGCCTGCCTGCACTTAGAGCAGTAGC    100

USP73p_001       100 GTTGGATTCCTGGATTGTTGTTCTAAACACACAGGGTTAAGCCTTCTTCC    149
                     ||||||||||||||||||||||||||||||||||||||||||||||||||
535194_001       101 GTTGGATTCCTGGATTGTTGTTCTAAACACACAGGGTTAAGCCTTCTTCC    150

USP73p_001       150 CTTGCAGAAAGATGTGTTGATCAATTCCACAAAA----------------    183
                     ||||||||||||||||||||||||||||||||||                
535194_001       151 CTTGCAGAAAGATGTGTTGATCAATTCCACAAAAAAAAAAAAAAAAAAAA    200

USP73p_001       184 ---    183
                        
535194_001       201 AAA    203
```

This features the the Poly(A) tail at the end of that nice match with the USP7-217 `3' downstream sequence`.

As for the part after the Poly(A) tail, that looks to mainly be adapter sequences, or something technical and not biological, because the main portion of it, `CGTTGATCGGAAGAGCGTCGTGTAGAGCAATACGTAACTGAA`, occurs a lot in the reads. Specifically, the Jupyter notebook [checking 3' end read_SRR23849628.535194 for adapter](checking_3prime_end_read_SRR23849628.535194_for_adapter.ipynb), shows it, or something close to it occurs, in 362 out of the 863 reads. And that wasn't even being thorough in looking into it.

So that accounts for the entire 895 bps of the read and contribures to the case supporting it being an instance of of the TSL4-rated transcript ENST00000567329 (USP7-217). This speaks to the power of Logan Search to help find things that may be rare.

With the features of that informative read dissected, let's consider it in more context.

**You may be saying but there was another read `fqgrep` identified...**

What about the other read `fqgrep` helped indicate has a match to the "retained intron" query sequence?  
At first glance it seems interesting that the other read is an even better match over the same 567 bp span:

```text
Alignment statistics for match #1
Score	Expect	Identities	Gaps	Strand
1017 bits(1127)	0.0	567/568(99%)	1/568(0%)	Plus/Plus
Query  349  GTTGGTGGAGCGATTACAAGAAGAGAAAAGGATCGAGGCTCAGAAGCGGAAGGAGCGGCA  408
            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
Sbjct  1    GTTGGTGGAGCGATTACAAGAAGAGAAAAGGATCGAGGCTCAGAAGCGGAAGGAGCGGCA  60

Query  409  GGAAGCCCATCTCTATATGCAAGTGCAGATAGTCGCAGAGGACCAGTTTTGTGGCCACCA  468
            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
Sbjct  61   GGAAGCCCATCTCTATATGCAAGTGCAGATAGTCGCAGAGGACCAGTTTTGTGGCCACCA  120

Query  469  AGGGAATGACATGTACGATGAAGAAAAAGTGAAATACACTGTGTTCAAAGTATTGAAGAA  528
            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
Sbjct  121  AGGGAATGACATGTACGATGAAGAAAAAGTGAAATACACTGTGTTCAAAGTATTGAAGAA  180

Query  529  CTCCTCGCTTGCTGAGTTTGTTCAGAGCCTCTCTCAGACCATGGGATTTCCACAAGATCA  588
            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
Sbjct  181  CTCCTCGCTTGCTGAGTTTGTTCAGAGCCTCTCTCAGACCATGGGATTTCCACAAGATCA  240

Query  589  AATTCGATTGTGGCCCATGCAAGCAAGGAGTAATGGAACAAAACGACCAGCAATGTTAGA  648
            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
Sbjct  241  AATTCGATTGTGGCCCATGCAAGCAAGGAGTAATGGAACAAAACGACCAGCAATGTTAGA  300

Query  649  TAATGAAGCCGACGGCAATAAAACAGTAAATATTGTTAATAGCGTATCTGGTTTGGAACC  708
            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
Sbjct  301  TAATGAAGCCGACGGCAATAAAACAGTAAATATTGTTAATAGCGTATCTGGTTTGGAACC  360

Query  709  GTGCAGAAGGCGTTAGTCCTCTGCACTTAGTGCAGCTTGTTTCCCTTTTGGTCCACTTAA  768
            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
Sbjct  361  GTGCAGAAGGCGTTAGTCCTCTGCACTTAGTGCAGCTTGTTTCCCTTTTGGTCCACTTAA  420

Query  769  CTAGAATTGGACGTTTTCTTCAATACTTGACTGTAGTTTTTCGCTCTGTCACCTAAGCCA  828
            ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
Sbjct  421  CTAGAATTGGACGTTTTCTTCAATACTTGACTGTAGTTTTTCGCTCTGTCACCTAAGCCA  480

Query  829  TTAGACTCTTCTAAAATCAGCGTCTCTTTTGGAAAATAGATGATTGAGCTCAGTGATAAT  888
            |||||||||||||||||||||||||||||||| |||||||||||||||||||||||||||
Sbjct  481  TTAGACTCTTCTAAAATCAGCGTCTCTTTTGG-AAATAGATGATTGAGCTCAGTGATAAT  539

Query  889  GAAAACCCTTGGACAATATTCCTGGAAA  916
            ||||||||||||||||||||||||||||
Sbjct  540  GAAAACCCTTGGACAATATTCCTGGAAA  567
```

However, it is larger than expected for USP7-217 and matches other parts of the unspliced USP7 gene as well, and so it may indeed be represenative of an unspliced transcript, or more likely, the result of contaminating DNA being sequenced.


That is all very insightful but there's 863 reads in `USP7_locus_mapped_SRX19662564_reads.fastq`. Can we scale the process of such analysis?

The rest of the notebook will be comprised of doing a rough sketch of that to classify the additional 861 reads in `USP7_locus_mapped_SRX19662564_reads.fastq`. Though we already have preliminary evidence from searching for the query pattern and biological variations of it, that it doesn't occur elsewhere, it would be nice to have independent evidence of that by looking at each sequence and considering it further with BLAST. USP7 has a lot of transcripts so we won't thoroughly classify all of them here, but we'll look for candidates that seem to stronly match the canonical USP7 (MANE) transcript, and look via alignment for any that, though they don't have a good match to the query sequence, may indicate a godd match to the TSL4-rated transcript ENST00000567329 (USP7-217) that retains the intron.

------

#### Combine in the result of examining for query sequence with BLAST to identify transcripts

Run the next cell to make databases based on the sequences of the transcripts, both the MANE Select one and the USP9-217 one.

In [11]:
!makeblastdb -in supporting_tables_and_data/Homo_sapiens_ENST00000344836_9_sequence_USP7-MANE_trancript.fasta -dbtype nucl
!makeblastdb -in supporting_tables_and_data/Homo_sapiens_ENST00000567329_1_sequence_USP7-217_trancript.fasta  -dbtype nucl



Building a new DB, current time: 04/17/2025 20:14:37
New DB name:   /home/jovyan/supporting_tables_and_data/Homo_sapiens_ENST00000344836_9_sequence_USP7-MANE_trancript.fasta
New DB title:  supporting_tables_and_data/Homo_sapiens_ENST00000344836_9_sequence_USP7-MANE_trancript.fasta
Sequence type: Nucleotide
Deleted existing Nucleotide BLAST database named /home/jovyan/supporting_tables_and_data/Homo_sapiens_ENST00000344836_9_sequence_USP7-MANE_trancript.fasta
Keep MBits: T
Maximum file size: 3000000000B
Adding sequences from FASTA; added 1 sequences in 0.000257015 seconds.




Building a new DB, current time: 04/17/2025 20:14:38
New DB name:   /home/jovyan/supporting_tables_and_data/Homo_sapiens_ENST00000567329_1_sequence_USP7-217_trancript.fasta
New DB title:  supporting_tables_and_data/Homo_sapiens_ENST00000567329_1_sequence_USP7-217_trancript.fasta
Sequence type: Nucleotide
Deleted existing Nucleotide BLAST database named /home/jovyan/supporting_tables_and_data/Homo_sapiens_ENST000

Now set to iterate on reads with BLAST query against each of the transcript options being screened for here. But also want to check for query sequence so know which ones have it and can use that in the programmatic screen.  
This is code towards doing that check for the query sequence:

In [12]:
import re
def parse_fqgrep_output_string_for_identitiers(fqgrep_output):
    """
    Parse fqgrep output string and extract the description lines (headers) from FASTQ sequences.
    
    Args:
        fqgrep_output: String containing fqgrep output
        
    Returns:
        A list of description lines.
    """
    # Pattern to match FASTQ header lines, which start with '@'
    # and are typically followed by a sequence identifier
    pattern = r'^@\S+$'
    # Find all matches in the content, using multiline mode
    description_lines = re.findall(pattern, fqgrep_output, re.MULTILINE)
    return description_lines

USP7_TSL4_RET_I = 'GGCAATAAAACAGTAAATATTGTTAATAGCG'
o = !fqgrep --reverse-complement --color auto {USP7_TSL4_RET_I} supporting_tables_and_data/USP7_locus_mapped_SRX19662564_reads.fastq
identifiers_from_reads_with_matches = parse_fqgrep_output_string_for_identitiers(o.n)
identifiers_from_reads_with_matches_without_ampersand = [x[1:]for x in identifiers_from_reads_with_matches]
#print(identifiers_from_reads_with_matches) # Uncomment for debugging
print(f"Screened reads for presence of match to `USP7_TSL4_RET_I` sequence:\nFound: {identifiers_from_reads_with_matches_without_ampersand}")

Screened reads for presence of match to `USP7_TSL4_RET_I` sequence:
Found: ['SRR23849628.5083175', 'SRR23849628.535194']


The next cell will actually iterate running the two BLAST jobs on each read to prepare for subsequenct classifying, it will take like 5 to 10 minutes to run the roughly 1700 BLAST comparisons (BE PATIENT!):

In [13]:
%%capture 
# ABOVE LINE to suppress a long stream of output as each sequence/ dataframe is processed
import os
from blast_to_df import blast_to_df
import pandas as pd


def process_fastq_to_individual_fastas_and_thenBLAST(fastq_file, output_dir="."):
    """
    Convert each read in a FASTQ file to an individual FASTA file,
    then some BLAST queries on each FASTA file to see what it matches 
    best out of the two transcripts. Or if matches well to neither.

    Save results in a list of two-entry tuples for now with each tuple 
    is the dataframe of results.
    
    Args:
        fastq_file (str): Path to input FASTQ file
        output_dir (str): Directory where to save individual FASTA files
    """
    max_reads=1800 # set low for development; set impossibly high for typical runs

    df_pairs_plus_fastafile = []
    
    # Create output directory if it doesn't exist
    os.makedirs(output_dir, exist_ok=True)
    
    with open(fastq_file, 'r') as fin:
        read_count = 0
        header = ""
        sequence = ""
        
        line_count = 0
        for line in fin:
            line = line.strip()
            line_count += 1
            mod = line_count % 4
            
            if mod == 1:  # Header line (starts with @)
                header = line[1:]  # Remove the '@' character
                read_count += 1
            elif mod == 2:  # Sequence line
                sequence = line
            elif mod == 0:  # Quality line (fourth line) - we've completed a read
                # Create individual FASTA file
                fasta_file = os.path.join(output_dir, f"{header}.fasta")
                
                with open(fasta_file, 'w') as fout:
                    fout.write(f">{header}\n{sequence}\n")
                
                # Now that the FASTA file exists, you can act on it
                results_tuple = process_individual_fasta_via_BLAST(fasta_file)
                # FOR DEVELOPMENT ONLY" check each has returned something. I was originally
                # planning to add a 'placeholder' / indicator for those not returning anything
                # good or dataframe without rows, but decided easy enough to keep checking for
                # later since would need to check for status either way,
                print(isinstance(results_tuple[0], pd.DataFrame) and len(results_tuple[0]) > 0) # Uncomment for development
                print(isinstance(results_tuple[1], pd.DataFrame)  and len(results_tuple[1]) > 0) # Uncomment for development
                
                df_pairs_plus_fastafile.append((results_tuple[0],results_tuple[1],fasta_file))
                              
        

                # Break after processing max_reads
                if read_count >= max_reads:
                    print(f"Reached maximum number of reads ({max_reads}). Stopping.")
                    break
    return df_pairs_plus_fastafile

import re
def parse_fqgrep_output_string_for_identitiers(fqgrep_output):
    """
    Parse fqgrep output string and extract the description lines (headers) from FASTQ sequences.
    
    Args:
        fqgrep_output: String containing fqgrep output
        
    Returns:
        A list of description lines.
    """
    # Pattern to match FASTQ header lines, which start with '@'
    # and are typically followed by a sequence identifier
    pattern = r'^@\S+$'
    # Find all matches in the content, using multiline mode
    description_lines = re.findall(pattern, fqgrep_output, re.MULTILINE)
    return description_lines

# as preparation for classification, also collect identifiers that have matches
# to the query sequence.
USP7_TSL4_RET_I = 'GGCAATAAAACAGTAAATATTGTTAATAGCG'
o = !fqgrep --reverse-complement --color auto {USP7_TSL4_RET_I} supporting_tables_and_data/USP7_locus_mapped_SRX19662564_reads.fastq
identifiers_from_reads_with_matches = parse_fqgrep_output_string_for_identitiers(o.n)
identifiers_from_reads_with_matches_without_ampersand = [x[1:]for x in identifiers_from_reads_with_matches]

def process_individual_fasta_via_BLAST(fasta_file):
    """
    Process an individual FASTA file to use BLAST to determine
    if matches MANE transcript or USP7-217 best.
    Return the dataframe pairs plus the FASTA file used in a 
    three member tuple
    
    Args:
        fasta_file (str): Path to FASTA file
    """
    print(f"Determining via BLAST best match: {fasta_file}")
    mane_result = !blastn -query {fasta_file} -db supporting_tables_and_data/Homo_sapiens_ENST00000344836_9_sequence_USP7-MANE_trancript.fasta -outfmt "6 qseqid sseqid stitle pident qcovs length mismatch gapopen qstart qend sstart send qframe sframe frames evalue bitscore qseq sseq" -task blastn
    the_217_result = !blastn -query {fasta_file} -db supporting_tables_and_data/Homo_sapiens_ENST00000567329_1_sequence_USP7-217_trancript.fasta  -outfmt "6 qseqid sseqid stitle pident qcovs length mismatch gapopen qstart qend sstart send qframe sframe frames evalue bitscore qseq sseq" -task blastn
    df_mane = blast_to_df(mane_result.n)
    df_217 = blast_to_df(the_217_result.n)
    #blast_df.head() # ONLY FOR DEBUGGING IN DEVELOPMENT
    
    # Process the newly made FASTA file
    # return the tuple of dfs
    return (df_mane,df_217)

# Example usage
#process_fastq_to_individual_fastas("input.fastq", "output_fastas")
blast_df_pairs_plus_file = process_fastq_to_individual_fastas_and_thenBLAST("supporting_tables_and_data/USP7_locus_mapped_SRX19662564_reads.fastq", "converted.fasta")

Now do the actual classification step:

In [14]:
%%capture
# classify each as match two either transcript or neither, 
# the ouput will be a four member tuple: 
#       -the max BLAST result
#       - the classification
#       - FASTA file
#       - length of query sequence
# Additional logic:
# if it has the RET_I 31-mer sequence it is going to get considered as 
# TSL4-rated USP-217 transcript version, as preliminary efforts covered
# above spelled that out. 
# Calling dataframes that have no values causes key errors, so there is a
# lot of checks early on that each dataframe, returned from BLAST search 
# against each of the two sequences, has content.
# This is only meant AS SIMPLE APPROXIMATION AT THIS TIME.
import os
from pyfaidx import Fasta
output_dir = "converted.fasta"
transcript_length = 567 #length of ENST00000567329 (USP7-217)

def get_fasta_sequence_length(fasta_file):
    """
    Get the length of a sequence in a FASTA file using pyfaidx.
    Assumes the FASTA file contains a single sequence.
    """
    sequence = Fasta(fasta_file)
    # Get the first (and only) sequence key
    first_sequence_key = list(sequence.keys())[0]
    sequence_length = len(sequence[first_sequence_key])
    return sequence_length
def check_if_matches_identified_sequences(FASTAfile_identifier):
    #identifier_from_FASTA_path = os.path.splitext(os.path.basename(FASTAfile_path))[0] #drop directory and `.fasta` extension
    # moved above conversion to main process calling this so could use in data collection
    print(FASTAfile_identifier)
    return FASTAfile_identifier in identifiers_from_reads_with_matches_without_ampersand

def classify_pair(df1,df2,FASTAfile):
    print(FASTAfile)
    fa_seq_length = get_fasta_sequence_length(FASTAfile)
    FASTAfile_identifier = os.path.splitext(os.path.basename(FASTAfile))[0] #drop directory and `.fasta` extension
    print(FASTAfile_identifier)
    print(f"{len(df1)} rows for MANE_df")
    print(f"{len(df2)} rows for Two17_df")
    
    current_sequence_contains_match_to_query_sequence = check_if_matches_identified_sequences(FASTAfile_identifier)

    # first make sure something reasonable.
    # Otherwise return 'neither'
    if not len(df1) > 0 and not len(df2) > 0:
        return 0,'neither',FASTAfile_identifier,fa_seq_length
    if len(df1) > 0:
        mane_length = df1.length[0]
        print(f"{mane_length} : MANE_df length column")
        if mane_length > 700 and len(df2) == 0:
            print(FASTAfile, "IS MANE!!!")
            return mane_length, "USP-MANE", FASTAfile_identifier,fa_seq_length
    if len(df2) > 0:
        two17_length = df2.length[0]
        print(f" {two17_length} : Two17_df length column")
    else:
        print("No content for Two17_df")
        two17_length = 0
        if mane_length > 300 and len(df2) == 0 and df1.qcovs[0] > 39:
            print(FASTAfile, "if Likely MANE!")
            return mane_length, "USP-MANE", FASTAfile_identifier,fa_seq_length
    print(f"{df1.qcovs[0]} : %query cover")
    
    # If has match the RET_I, consider it as USP-217 given preliminary
    # results here. Unless very long and then it probably 'unspliced/
    # genomic'.(Obviously, this may need customization if adapting to 
    # different query sequences and data.)
    if current_sequence_contains_match_to_query_sequence:
        if fa_seq_length < (transcript_length * 2):
            return two17_length, "USP-217", FASTAfile_identifier,fa_seq_length
        else:
            return two17_length, "unspliced/genomic", FASTAfile_identifier,fa_seq_length
    
    # Continue to make sure something reasonable relative each other.
    # Otherwise return 'neither'
    if df1.qcovs[0] < 40:
        return df1.length[0],'neither',FASTAfile_identifier,fa_seq_length

    if mane_length >= two17_length:
        return mane_length, "USP-MANE", FASTAfile_identifier,fa_seq_length  # First dataframe has larger or equal value
    else:
        # based on my searches with fuzzy version of RET_I 31-mer sequence, I don't 
        # expect there to be good matches without that but it is here in case it
        # catches something and also to more easily set-up for adapting to different
        # query and data.
        return two17_length, "USP-217", FASTAfile_identifier,fa_seq_length  # Second dataframe has larger value for length
    return df1.length[0],'neither',FASTAfile_identifier,fa_seq_length
classifications_list = []        
for mane_df,two17_df,FASTAfile in blast_df_pairs_plus_file:
    returned_four_item_tuple = classify_pair(mane_df,two17_df, FASTAfile)
    print(returned_four_item_tuple)
    classifications_list.append(returned_four_item_tuple)

In [15]:
classifications_df = pd.DataFrame(classifications_list, columns=['best_hit_length', 'classification', 'read_id', 'read_length'])
classifications_df

best_hit_length classification              read_id  read_length
0                106        neither  SRR23849628.2721367         2012
1                 17        neither  SRR23849628.1423419         2829
2                 20        neither  SRR23849628.3205402         1784
3                 16        neither   SRR23849628.870297         3009
4                105        neither  SRR23849628.2788216         2288
..               ...            ...                  ...          ...
858               12        neither  SRR23849628.1331608          959
859               11        neither  SRR23849628.2284886          924
860               18        neither  SRR23849628.1243519         2486
861                0        neither   SRR23849628.524796          832
862               18        neither  SRR23849628.6441316         1220

[863 rows x 4 columns]

With all 863 classified, ready to see how the classifications break down:

In [16]:
classification_counts = classifications_df['classification'].value_counts()
print(classification_counts)

classification
USP-MANE             556
neither              305
unspliced/genomic      1
USP-217                1
Name: count, dtype: int64


So ENST00000567329 (USP7-217) is like 1 to 556 of the other USP-7 transcripts.

Not even quite 0.2% of all identified USP7 transcripts.  
Probably explains in part why it is classified as TSL4.

More about those results:  
Also, it is important to note the BLAST comparisons didn't turn up any additional TSL4-rated transcript ENST00000567329 (USP7-217) candidates.

Keep in mind that this is a rough sketch at this time. Several classified in this simple scheme as 'MANE' will actually be unspliced/genomic, too. Plus, I'm sure some classified as 'neither' are some of the other USP7 transcripts Ensembl lists in the transcript table, but I haven't included a way to sort that for those that don't have a match to the RET_I 31-mer sequence. In other words, there's lots of other transcripts in this region, and so the process could be made to be much better at classifying. 
One could envision checking for splice junctions or lack thereof. Plus checking what positions in expected transcripts are spanned by the reads.    
However, this  should give a good general idea of the rough number of USP7 transcripts that aren't TSL4-rated USP7-217 that has the retained intron.

I also would be remiss not pointing out the 0.2% may be a gross overstimate. We only have one example in the one dataset. That is no way to assess. It may be pure chance that one example showed up in the 6.4 million reads. With only one it is hard to take an average. This is a classic case of a limited sample. Not to mention, there may be related biosamples that don't have it and indicate it is more rare. We can easily see why it is TSL4-rated after this and how Logan Search can be helpful if you want to go digging in the Sequence Read Archive (SRA) for evidence of your transcript of interest.

I also think the results presented here suggest a lot of the hits to the query for the retained intron sequence could also be contamination or unspliced transcripts. If I had to do this all over again, I'd choose to make this demonstration for a transcript with an ususual splicing event and use that as a basis to investigate. That way I can look for evidence of the unsual splice junction and the possibility of genomic contamination or unspliced transcripts would not be a confounding factor. Also, I'd probably come up with a combination of scoring based on presence of an exact or similar to query sequences alone with assessment of the BLAST hits, taking into account the spanned positions, all combined to indicate a score. And then use that score to decide classification of each read. That being said, this endeavor parallels one plenty of researchers may face, and the fact this endeavor still worked to find an example of this poorly supported transcript speaks to the power of Logan Search.

-----

Enjoy!